deepchem/feat/dft_featurizer.py

```diff
@@ -0,0 +1,332 @@
+            mol = Mol(**self["kwargs"])
+            return mol
+        else:
+            raise RuntimeError("Unknown system type: %s" % systype)
+
+    def get_cache(self, s: str):
+        return self._caches.get(s, None)
+
+    def set_cache(self, s: str, obj) -> None:
+        self._caches[s] = obj
+
+class Entry(dict):
+    """
+    Interface to the entry of the dataset.
+    Entry class should not be initialized directly, but created through
+    ``Entry.create``
+    """
+
+    created_entries: Dict[str, Entry] = {}
+
+    @classmethod
+    def create(
+        cls,
+        entry_dct: Union[Dict, Entry],
+        device: torch.device,
+        dtype: torch.dtype = torch.double
+    ) -> Entry:
+        if isinstance(entry_dct, Entry):
+            # TODO: should we add dtype and device checks here?
+            return entry_dct
+
+        s = str(entry_dct)
+        if s not in cls.created_entries:
+            tpe = entry_dct["type"]
+            kwargs = {
+                "entry_dct": entry_dct,
+                "dtype": dtype,
+                "device": device,
+            }
+            obj = {
+                "ae": EntryAE,
+                "ie": EntryIE,
+                "dm": EntryDM,
+                "dens": EntryDens,
+                "force": EntryForce,
+            }[tpe](**kwargs)
+            cls.created_entries[s] = obj
+        return cls.created_entries[s]
+
+    def __init__(self,
+                 entry_dct: Dict,
+                 device: torch.device,
+                 dtype: torch.dtype = torch.double):
+        super().__init__(entry_dct)
+        self._systems = [System.create(p) for p in entry_dct["systems"]]
+        self._dtype = dtype
+        self._device = device
+
+        self._trueval_is_set = False
+        self._trueval = torch.tensor(0.0).to(device)
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self._dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self._device
+
+    def get_systems(self) -> List[System]:
+        """
+        Returns the list of systems in the entry
+        """
+        return self._systems
+
+    @abstractproperty
+    def entry_type(self) -> str:
+        """
+        Returning the type of the entry of the dataset
+        """
+        pass
+
+    def get_true_val(self) -> torch.Tensor:
+        if not self._trueval_is_set:
+            self._trueval = self._get_true_val()
+            self._trueval_is_set = True
+        return self._trueval
+
+    @abstractmethod
+    def _get_true_val(self) -> torch.Tensor:
+        """
+        Get the true value of the entry.
+        """
+        pass
+
+    @abstractmethod
+    def get_val(self, qcs: List[BaseKSCalc]) -> torch.Tensor:
+        """
+        Calculate the value of the entry given post-run QC objects.
+        """
+        pass
+
+    @abstractmethod
+    def get_loss(self, val: torch.Tensor,
+                 true_val: torch.Tensor) -> torch.Tensor:
```

Comment:
Why does an entry object know about the pytorch loss? We may have to refactor this class a bit since it feels like a leaky abstraction